gulpfile.js

```diff
@@ -1188,7 +1188,16 @@ function generateTypeScriptDefinitions(
       (match, p1) => `= WebGLConstants.${p1}`
     )
     // Strip const enums which can cause errors - https://www.typescriptlang.org/docs/handbook/enums.html#const-enum-pitfalls
-    .replace(/^(\s*)(export )?const enum (\S+) {(\s*)$/gm, "$1$2enum $3 {$4");
+    .replace(/^(\s*)(export )?const enum (\S+) {(\s*)$/gm, "$1$2enum $3 {$4")
+    // Replace JSDoc generation version of defined with an improved version using TS type predicates
+    .replace(
+      /defined\(value: any\): boolean/gm,
+      "defined<Type>(value: Type): value is NonNullable<Type>"
+    )
+    .replace(
+      /\/\*\*[\*\s\w]*?\*\/\nexport const Check: any;/m,
+      `\n${readFileSync("./packages/engine/Source/Core/Check.d.ts").toString()}`
```

Comment:
Can this be made generic to and `**/Source/**/*.d.ts` file? That way any future additions don't need to tweak build code.

Reply:
I would love to do that but I couldn't find or think of a nice clean way to do that. As it is it needs to be specific to replace the definition in the types file for the corresponding code. Otherwise we'd end up with a duplicated definition which breaks TS. I'd love if `tsd-jsdoc` let you pass in a set of types files to use in places like this.